lib/argonaut/bot.ex

```diff
@@ -0,0 +1,68 @@
+defmodule Bot do
+  alias Argonaut.Reservations
+
+  @env_app_re "(?<env>.*):(?<app>.*)"
+  # handle im, i'm, i am
+  @i_am_re "I(?:'m|m| am)"
+  # handle whos, who's, who is
+  @who_is_re "Who(?:'s|s| is)"
+
+  @matchers [
+    {:reservation_action, ~r/#{@env_app_re} (?<action>info|release|renew|reserve)/},
+    {:reservation_info, ~r/#{@who_is_re} using #{@env_app_re}(\?)?/},
+    {:reservation_release, ~r/#{@i_am_re} done using #{@env_app_re}/},
+    {:reservation_renew, ~r/#{@i_am_re} still using #{@env_app_re}/},
+    {:reservation_reserve, ~r/#{@i_am_re} using #{@env_app_re}(for (?<reason>.*))?/}
+  ]
+
+  @default_handler {:default_handler, nil}
+
+  def reply(%{message: message, user: user}) do
+    {action, regex} =
+      Enum.find(@matchers, @default_handler, fn {_, re} -> Regex.match?(re, message) end)
+
+    if action == :default_handler do
+      apply(__MODULE__, action, [])
+    else
+      matches = Regex.named_captures(regex, message)
+      apply(__MODULE__, action, [matches, user])
+    end
+  end
+
+  def reservation_action(%{"action" => action, "app" => app, "env" => env} = params, user) do
+    action_function = String.to_atom("reservation_#{action}")
+
+    apply(__MODULE__, action_function, [params, user])
```

Comment:
This is a fairly large vulnerability.
1) it's a DOS vector because you're dynamically creating atoms based on user input and atoms are not garbage collected.
2) You're taking that dynamically created atom and feeding it to `:erlang.apply`. That's just asking for trouble.

Reply:
Someone else had the same opinion... I will fix that.

Taking a step back though, would this be mitigated by ensuring the length of message to be <= 512 chars?